### PR TITLE
DIV-4879 - initial work of changing tabs to match stages

### PR DESCRIPTION
**Pull request by hosainnet**
- [x] commit messages are meaningful and follow good commit message guidelines
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###

DIV-4879

### Change description ###

In summary, replacing the Petitioner, Respondent, Reason for Divorce etc tabs with
Petition, AOS, Decree Nisi, Outcome of DN, Decree Absolute 

i.e each tab contains the fields which were set by the particular stage of the divorce, not by the "party"

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: definitions/divorce/json/CaseTypeTab.json [hunk 1]
```diff
@@ -13,68 +13,68 @@
     "LiveFrom": "01/01/2017",
     "CaseTypeID": "DIVORCE",
     "Channel": "CaseWorker",
-    "TabID": "solicitorDetails",
-    "TabLabel": "Solicitor",
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: Moved Solicitor tab to Petition

File: definitions/divorce/json/CaseTypeTab.json [hunk 2]
```diff
@@ -83,99 +83,99 @@
     "LiveFrom": "01/01/2017",
     "CaseTypeID": "DIVORCE",
     "Channel": "CaseWorker",
-    "TabID": "petitionerDetails",
-    "TabLabel": "Petitioner",
-    "TabDisplayOrder": 3,
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: Renamed `petitionerDetails` -> `petitionDetails` and `Petitioner` -> `Petition`

File: definitions/divorce/json/CaseTypeTab.json [hunk 3]
```diff
@@ -250,957 +250,897 @@
...
   },
   {
     "LiveFrom": "01/01/2017",
     "CaseTypeID": "DIVORCE",
     "Channel": "CaseWorker",
-    "TabID": "petitionerDetails",
-    "TabLabel": "Petitioner",
-    "TabDisplayOrder": 3,
-    "CaseFieldID": "labelDNAnswers",
-    "TabFieldDisplayOrder": 31
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: Removed unwanted label

File: definitions/divorce/json/CaseTypeTab.json [hunk 4]
```diff
@@ -250,957 +250,897 @@
...
-    "CaseTypeID": "DIVORCE",
-    "Channel": "CaseWorker",
-    "TabID": "petitionerDetails",
-    "TabLabel": "Petitioner",
-    "TabDisplayOrder": 3,
+    "TabID": "dnDetails",
+    "TabLabel": "Decree Nisi",
+    "TabDisplayOrder": 4,
     "CaseFieldID": "PermittedDecreeNisiReason",
-    "TabFieldDisplayOrder": 32
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: DN fields to a new DN tab

File: definitions/divorce/json/CaseTypeTab.json [hunk 5]
```diff
@@ -250,957 +250,897 @@
...
-  },
-  {
-    "LiveFrom": "01/01/2019",
-    "CaseTypeID": "DIVORCE",
-    "Channel": "CaseWorker",
-    "TabID": "petitionerDetails",
-    "TabLabel": "Petitioner",
-    "TabDisplayOrder": 3,
-    "CaseFieldID": "labelOutcomeofDN",
-    "TabFieldDisplayOrder": 54
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: Removed unwanted label

File: definitions/divorce/json/CaseTypeTab.json [hunk 6]
```diff
@@ -250,957 +250,897 @@
...
     "LiveFrom": "01/01/2017",
     "CaseTypeID": "DIVORCE",
     "Channel": "CaseWorker",
-    "TabID": "petitionerDetails",
-    "TabLabel": "Petitioner",
+    "TabID": "outcomeOfDnDetails",
+    "TabLabel": "Outcome of Decree Nisi",
     "TabDisplayOrder": 3,
     "CaseFieldID": "DNApprovalDate",
-    "TabFieldDisplayOrder": 55
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: Outcome of DN to new Outcome of DN tab

File: definitions/divorce/json/CaseTypeTab.json [hunk 7]
```diff
@@ -250,957 +250,897 @@
...
-  },
-  {
-    "LiveFrom": "01/01/2017",
-    "CaseTypeID": "DIVORCE",
-    "Channel": "CaseWorker",
-    "TabID": "respondentDetails",
-    "TabLabel": "Respondent",
-    "TabDisplayOrder": 4,
-    "CaseFieldID": "createdDate",
-    "TabFieldDisplayOrder": 2
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: `D8caseReference` and `createdDate` are petitioner fields so removed them from AOS

File: definitions/divorce/json/CaseTypeTab.json [hunk 8]
```diff
@@ -250,957 +250,897 @@
...
     "CaseTypeID": "DIVORCE",
     "Channel": "CaseWorker",
-    "TabID": "respondentDetails",
-    "TabLabel": "Respondent",
-    "TabDisplayOrder": 4,
+    "TabID": "petitionDetails",
+    "TabLabel": "Petition",
+    "TabDisplayOrder": 2,
     "CaseFieldID": "D8RespondentFirstName",
-    "TabFieldDisplayOrder": 3
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: `D8RespondentFirstName` fields are generally Petition fields

File: definitions/divorce/json/CaseTypeTab.json [hunk 9]
```diff
@@ -250,957 +250,897 @@
...
     "Channel": "CaseWorker",
-    "TabID": "respondentDetails",
-    "TabLabel": "Respondent",
-    "TabDisplayOrder": 4,
+    "TabID": "aosDetails",
+    "TabLabel": "AOS",
+    "TabDisplayOrder": 3,
     "CaseFieldID": "RespNameDifferentToMarriageCertExplain",
-    "TabFieldDisplayOrder": 6
+    "TabFieldDisplayOrder": 1
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: New AOS tab

File: definitions/divorce/json/CaseTypeTab.json [hunk 10]
```diff
@@ -13,68 +13,68 @@
...
     "CaseTypeID": "DIVORCE",
     "Channel": "CaseWorker",
-    "TabID": "solicitorDetails",
-    "TabLabel": "Solicitor",
+    "TabID": "petitionDetails",
+    "TabLabel": "Petition",
     "TabDisplayOrder": 2,
     "CaseFieldID": "PetitionerSolicitorName",
-    "TabFieldDisplayOrder": 1
+    "TabFieldDisplayOrder": 22
```
Comment (hosainnet) [inline definitions/divorce/json/CaseTypeTab.json]: Ordering of fields by TabFieldDisplayOrder is a bit crap for now, given this is a large PR I'll try to do it after this is merged

**Comment by qzhou-hmcts** [comment]
Also bump up version before merge

**Comment by hosainnet** [comment]
Yep needs a thorough QA

File: test/definitions/divorce/CaseTypeTab.test.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    it('should contain a proper sequence for TabFieldDisplayOrder with no gaps', function() {
+      const tabIds = _.uniq(_.map(caseTypeTab, 'TabID'));
+      tabIds.forEach(tabId => {
+        const allFieldsPerTab = _.filter(caseTypeTab, (field) => {
+          return field.TabID === tabId;
+        });
+        const allTabFieldDisplayOrderNumbers = _.map(allFieldsPerTab, (field) => {
+          return field.TabFieldDisplayOrder;
+        }).sort((a, b) => a - b);
+        for(var i = 1; i < allTabFieldDisplayOrderNumbers.length; i++) {
```
Comment (alexwakeman) [inline test/definitions/divorce/CaseTypeTab.test.js]: Minor - should we have es-lint involved now? I'm not a fan of no spacing `for(var`.
Also we should be using `let` or `const` not `var`.

File: test/definitions/divorce/CaseTypeTab.test.js [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            field1.TabFieldDisplayOrder === field2.TabFieldDisplayOrder
+        );
+        expect(uniqResults).to.eql(allFieldsPerTab);
+      });
+    });
+  });
+
+  describe('sequence', function() {
+    it('should contain a proper sequence for TabFieldDisplayOrder with no gaps', function() {
+      const tabIds = _.uniq(_.map(caseTypeTab, 'TabID'));
```
Comment (alexwakeman) [inline test/definitions/divorce/CaseTypeTab.test.js]: We could make this a bit tidier by importing only the lodash functions we need:
`import { uniq, map } from 'lodash';`
```
const tabIds = uniq(map(caseTypeTab, 'TabID'));
```

File: test/definitions/divorce/CaseTypeTab.test.js [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+      const uniqResult = _.uniqWith(
+        caseTypeTab,
+        (field1, field2) =>
+          field1.TabID === field2.TabID &&
+          field1.CaseFieldID === field2.CaseFieldID
+      );
+      expect(uniqResult).to.eql(caseTypeTab);
+    });
+    it('should contain a unique tab field display order ID field tab ID (no duplicate field order in a tab)', function() {
+      const tabIds = _.uniq(_.map(caseTypeTab, 'TabID'));
```
Comment (alexwakeman) [inline test/definitions/divorce/CaseTypeTab.test.js]: `TabID` could should be a re-usable external constant.

File: test/definitions/divorce/CaseTypeTab.test.js [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+      tabIds.forEach(tabId => {
+        const allFieldsPerTab = _.filter(caseTypeTab, (field) => {
+          return field.TabID === tabId;
+        });
```
Comment (alexwakeman) [inline test/definitions/divorce/CaseTypeTab.test.js]:
```
const allFieldsPerTab = _.filter(caseTypeTab, field => field.TabID === tabId);
```

File: test/definitions/divorce/CaseTypeTab.test.js [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        expect(allTabOrders).to.eql([expected[tabId]]);
+      });
+    });
+  });
+
+  describe('validation', function() {
+    it('should contain a valid Tab IDs', function() {
+      const tabIds = _.uniq(_.map(caseTypeTab, 'TabID'));
+      expect(tabIds).to.eql([
+        "History",
```
Comment (alexwakeman) [inline test/definitions/divorce/CaseTypeTab.test.js]: Mixing double-quotes with single quotes for strings - again es-lint might be good?

File: test/definitions/divorce/CaseTypeTab.test.js [hunk 16]
```diff
@@ -0,0 +1,112 @@
+const expect = require('chai').expect;
+const assert = require('chai').assert;
+const _ = require('loadsh');
+const caseTypeTab = require('../../../definitions/divorce/json/CaseTypeTab');
```
Comment (alexwakeman) [inline test/definitions/divorce/CaseTypeTab.test.js]: Might be a good idea to clone this to keep the original data immutable.

File: test/definitions/divorce/CaseTypeTab.test.js (hunk 11 above)
Comment (hosainnet) [reply test/definitions/divorce/CaseTypeTab.test.js]: Done 👍

File: test/definitions/divorce/CaseTypeTab.test.js (hunk 12 above)
Comment (hosainnet) [reply test/definitions/divorce/CaseTypeTab.test.js]: Done 👍

File: test/definitions/divorce/CaseTypeTab.test.js (hunk 13 above)
Comment (hosainnet) [reply test/definitions/divorce/CaseTypeTab.test.js]: Done 👍

File: test/definitions/divorce/CaseTypeTab.test.js (hunk 14 above)
Comment (hosainnet) [reply test/definitions/divorce/CaseTypeTab.test.js]: I tried this, but with `eslint --fix` it keeps reverting it to the longer format, any idea?

File: test/definitions/divorce/CaseTypeTab.test.js (hunk 15 above)
Comment (hosainnet) [reply test/definitions/divorce/CaseTypeTab.test.js]: Done 👍

File: test/definitions/divorce/CaseTypeTab.test.js (hunk 16 above)
Comment (hosainnet) [reply test/definitions/divorce/CaseTypeTab.test.js]: How to?

Comment (alexwakeman) [reply test/definitions/divorce/CaseTypeTab.test.js]:
```
const caseTypeTab = Object.assign({}, require('../../../definitions/divorce/json/CaseTypeTab'));
```

File: test/definitions/divorce/CaseTypeTab.test.js (hunk 14 above)
Comment (alexwakeman) [reply test/definitions/divorce/CaseTypeTab.test.js]: We can use an es-lint override to allow this but it's up to you

File: test/definitions/divorce/CaseTypeTab.test.js (hunk 16 above)
Comment (hosainnet) [reply test/definitions/divorce/CaseTypeTab.test.js]: 👍

**Comment by hosainnet** [comment]
Verified basic scenarios on Demo with @nsingamsetty - will keep an eye on AAT

**Pull request merged**